state/pool.go

```diff
@@ -37,19 +48,78 @@ func (p *StatePool) Get(modelUUID string) (*State, error) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	st, ok := p.pool[modelUUID]
+	item, ok := p.pool[modelUUID]
+	if ok && item.remove {
+		// We don't want to allow increasing the refcount of a model
+		// that's been removed.
+		return nil, errors.Errorf("model %v has been removed", modelUUID)
+	}
 	if ok {
-		return st, nil
+		item.references++
+		return item.state, nil
 	}
 
 	st, err := p.systemState.ForModel(names.NewModelTag(modelUUID))
 	if err != nil {
 		return nil, errors.Annotatef(err, "failed to create state for model %v", modelUUID)
 	}
-	p.pool[modelUUID] = st
+	p.pool[modelUUID] = &PoolItem{state: st, references: 1}
 	return st, nil
 }
 
+// Put indicates that the client has finished using the State. If the
+// state has been marked for removal, it will be closed and removed
+// when the final Put is done.
+func (p *StatePool) Put(modelUUID string) error {
+	if modelUUID == p.systemState.ModelUUID() {
+		// We don't maintain a refcount for the controller.
+		return nil
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	item, ok := p.pool[modelUUID]
+	if !ok {
+		return errors.Errorf("unable to return unknown model %v to the pool", modelUUID)
+	}
+	if item.references == 0 {
+		return errors.Errorf("state pool refcount for model %v is already 0", modelUUID)
+	}
+	item.references--
+	return p.maybeRemoveItem(modelUUID, item)
+}
+
+// Remove takes the state out of the pool and closes it, or marks it
+// for removal if it's currently being used (indicated by Gets without
+// corresponding Puts).
+func (p *StatePool) Remove(modelUUID string) error {
+	if modelUUID == p.systemState.ModelUUID() {
+		// We don't manage the controller state.
+		return nil
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	item, ok := p.pool[modelUUID]
+	if !ok {
+		// Don't require the client to keep track of what we've seen -
+		// ignore unknown model ids.
```

Comment:
I mean, they *are* identifiers (which happen to be universally unique). I'm not sure every mention of them needs to highlight the universal part, given that people aren't digging around inside them. But I don't feel strongly about this, so I don't mind changing it if other people feel differently.